sinker/sinker.go

```diff
@@ -118,33 +125,47 @@ func (s *SQLSinker) HandleBlockScopedData(ctx context.Context, data *pbsubstream
 		return fmt.Errorf("apply database changes: %w", err)
 	}
 
+	s.lastSeenCursor = cursor
+	s.lastSeenFinalBlock = data.FinalBlockHeight
```

Comment:
Should we simply add those to `flush` directly instead? That would avoid maintaining some state before calling the function.

It also makes me wondering what happen if `flush` is called through `OnTerminating` but last block cursor/final block is the one before. It seems that it's a possibility here.

Reply:
> Should we simply add those to flush directly instead? That would avoid maintaining some state before calling the function.

That doesn't work because we don't know the last seen cursor and final block when executing `flush` from `OnTerminating`

> It also makes me wondering what happen if flush is called through OnTerminating but last block cursor/final block is the one before. It seems that it's a possibility here.

True, I guess we can just lock a mutex in `HandleBlockScopedData` to ensure that we always have a block fully processed before anything is flushed.